File: typescript/src/document.ts
import type { Document, DocumentSummary, UnlockResult } from "./model.js";
import { decodeBase64Url, encodeHex } from "./encoding.js";
import { validateDocument, parseDocument } from "./validation.js";
import { publicKeyFromPrivate, publicKeyMatches } from "./secp256k1.js";
import { derivePassword, saltFromDocument } from "./password.js";
import { decryptAesGcm } from "./aesGcm.js";
import { keyHoldError } from "./errors.js";

export { parseDocument };

export function serializeDocument(document: Document): string {
  return JSON.stringify(validateDocument(document));
}
export function summary(document: Document): DocumentSummary {
  const value = validateDocument(document);
  return {
    format: value.format,
    version: value.version,
    label: value.label,
    publicKeyHex: value.publicKeyHex,
  };
}

export async function unlockDocument(
  document: Document,
  password: string,
): Promise<UnlockResult> {
  const value = validateDocument(document);
  try {
    const key = await derivePassword(
      password,
      saltFromDocument(value.keyDerivation.saltB64Url),
      value.keyDerivation,
    );
    const plaintext = await decryptAesGcm(
      key,
      decodeBase64Url(value.cipher.ivB64Url),
      decodeBase64Url(value.cipher.ciphertextAndTagB64Url),
    );
    if (!publicKeyMatches(plaintext, value.publicKeyHex))
      throw keyHoldError(
        "unlock_failed",
        "private key does not match document public key",
      );
    return {
      privateKey: plaintext,
      publicKeyHex: encodeHex(publicKeyFromPrivate(plaintext)),
    };
  } catch (cause) {
    throw keyHoldError("unlock_failed", "unable to unlock document", cause);
  }
}

export const parse = parseDocument;
export const validate = validateDocument;
export const serialize = serializeDocument;
export const unlock = unlockDocument;

File: go/document.go
package keyhold

import (
	"encoding/json"
	"unicode/utf8"
)

// Serialize validates and encodes a document as canonical JSON.
func Serialize(document Document) ([]byte, error) {
	if err := Validate(document); err != nil {
		return nil, err
	}
	return json.Marshal(document)
}

// Summary returns the non-secret metadata of a validated document.
func Summary(document Document) (DocumentSummary, error) {
	if err := Validate(document); err != nil {
		return DocumentSummary{}, err
	}
	return DocumentSummary{Format: document.Format, Version: document.Version, Label: document.Label, PublicKeyHex: document.PublicKeyHex}, nil
}

func unlockResult(document Document, plaintext []byte) (*UnlockResult, error) {
	publicKey, err := PublicKeyFromPrivate(plaintext)
	if err != nil || EncodeHex(publicKey) != document.PublicKeyHex {
		return nil, E(ErrUnlockFailed, "private key does not match document public key", err)
	}
	return &UnlockResult{PrivateKey: plaintext, PublicKeyHex: EncodeHex(publicKey)}, nil
}

// Unlock decrypts a document and verifies that the recovered private key matches its public key.
func Unlock(document Document, password string) (*UnlockResult, error) {
	if err := Validate(document); err != nil {
		return nil, err
	}
	if password == "" || !utf8.ValidString(password) {
		return nil, E(ErrUnlockFailed, "unable to unlock document", nil)
	}
	salt, err := DecodeBase64URL(document.KeyDerivation.SaltB64Url, "saltB64Url")
	if err != nil {
		return nil, E(ErrUnlockFailed, "unable to unlock document", err)
	}
	key := derivePassword(password, salt, document.KeyDerivation.Iterations)
	iv, err := DecodeBase64URL(document.Cipher.IVB64Url, "ivB64Url")
	if err != nil {
		return nil, E(ErrUnlockFailed, "unable to unlock document", err)
	}
	joined, err := DecodeBase64URL(document.Cipher.CiphertextAndTagB64Url, "ciphertextAndTagB64Url")
	if err != nil {
		return nil, E(ErrUnlockFailed, "unable to unlock document", err)
	}
	plaintext, err := decryptAESGCM(key, iv, joined)
	if err != nil {
		return nil, E(ErrUnlockFailed, "unable to unlock document", err)
	}
	result, err := unlockResult(document, plaintext)
	if err != nil {
		return nil, E(ErrUnlockFailed, "unable to unlock document", err)
	}
	return result, nil
}
